File: MobileApp/CareMeApp/NavigationDrawer/GCM/RegistrationIntentService.cs
using Android.App;
using Android.Preferences;
using Android.Gms.Iid;
using Android;
using Android.Gms.Gcm;
using Android.Util;
using Android.Support.V4.Content;
using System;
using Android.Content;
using NavigationDrawer;
using edu.uncc.homework4.Models;
using System.Net.Http;
using Android.Widget;

namespace edu.uncc.homework4
{
	[Service (Exported = false)]
	public class RegistrationIntentService : IntentService
	{
		const string TAG = "RegIntentService";
		static readonly string[] TOPICS = {"global"};
        ISharedPreferences prefs;

		public RegistrationIntentService() : base (TAG) {
           // prefs = PreferenceManager.GetDefaultSharedPreferences(this);
        }

		protected override void OnHandleIntent (Intent intent)
		{
            prefs = PreferenceManager.GetDefaultSharedPreferences(this);

			try {
				lock (TAG) {
					var instanceID = InstanceID.GetInstance(this);
					var token = instanceID.GetToken(GetString(NavigationDrawer.Resource.String.gcm_defaultSenderId),
						GoogleCloudMessaging.InstanceIdScope, null);
					Log.Info(TAG, "GCM Registration Token: " + token);
					SendRegistrationToServer(token);
                    //SubscribeTopics(token);
                    prefs.Edit().PutBoolean(CONSTANTS.RECEIVE_NOTIFICATION_BOOLEAN, true).Apply();
				}
			} catch (Exception e) {
				Log.Debug(TAG, "Failed to complete token refresh", e);
                prefs.Edit().PutBoolean(CONSTANTS.RECEIVE_NOTIFICATION_BOOLEAN, false).Apply();
			}
			var registrationComplete = new Intent(CONSTANTS.REGISTRATION_COMPLETE);
			LocalBroadcastManager.GetInstance(this).SendBroadcast(registrationComplete);
		}

		void SendRegistrationToServer(string token) {
            // Add custom implementation, as needed.
            string deviceId = prefs.GetString(Constants.DEVICEID, "");

            if (!deviceId.Equals(token))
            {
                var model = new DeviceIdModel
                {
                    UserId = prefs.GetString(CONSTANTS.USERID,""),
                    DeviceId = token
                };

                try
                {
                    var client = new HttpClient();
                    client.DefaultRequestHeaders.Add("Authorization", "Bearer " + prefs.GetString(CONSTANTS.AUTH_HEADER, ""));
                    client.PostAsync(CONSTANTS.POST_DEVICEID_URL, new FormUrlEncodedContent(model.ToDict())).ContinueWith((response) => {

                        Console.Write(response);
                    });
                    
                }catch(Exception e)
                {
                    Toast.MakeText(this, e.Message, ToastLength.Short).Show();

                }               
                    
                //post
                Console.Write("token is " + token);
            }
        }

		void SubscribeTopics(string token) {
			foreach (var topic in TOPICS) {
				var pubSub = GcmPubSub.GetInstance(this);
				pubSub.Subscribe(token, "/topics/" + topic, null);
			}
		}
	}
}

File: MobileApp/MyApplication/MyApplication/app/src/main/java/edu/uncc/homework4/RegistrationIntentService.java
package edu.uncc.homework4;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.util.Log;

import com.google.android.gms.gcm.GoogleCloudMessaging;
import com.google.android.gms.iid.InstanceID;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

/**
 * Created by Nitin on 11/19/2017.
 */

public class RegistrationIntentService extends IntentService {
    SharedPreferences prefs;

    String access_token, user_id;
    public RegistrationIntentService(){
        super("");
    }

    /**
     * Creates an IntentService.  Invoked by your subclass's constructor.
     *
     * @param name Used to name the worker thread, important only for debugging.
     */
    public RegistrationIntentService(String name) {
        super(name);
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {

        prefs = getApplicationContext().getSharedPreferences(Constants.PREFS, Context.MODE_PRIVATE);
        final SharedPreferences.Editor editor = prefs.edit();

        access_token = prefs.getString(Constants.AUTH_HEADER,"");
        user_id = prefs.getString(Constants.USERID,"");

        InstanceID instanceID = InstanceID.getInstance(this);
        String token = null;

        try {
            token = instanceID.getToken(getApplicationContext().getString(R.string.gcm_defaultSenderId),
                    GoogleCloudMessaging.INSTANCE_ID_SCOPE, null);
            editor.putString(Constants.RECEIVE_NOTIFICATION_BOOLEAN,"Yes");
            editor.commit();
        } catch (IOException e) {
            e.printStackTrace();
        }

        sendDeviceID(token, user_id);
    }

    public void sendDeviceID(String device_id, String user_id){

        RequestBody formBody = new FormBody.Builder()
                .add("UserId", user_id)
                .add("DeviceId", device_id)
                .build();

        Request request = new Request.Builder()
                .url(Constants.POST_DEVICEID_URL)
                .header("Authorization", "Bearer "+ access_token)
                .post(formBody)
                .build();

        OkHttpClient client = new OkHttpClient();

        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {

            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                Log.d("demo","refresh token " + response.body().string() );
            }
        });
    }
}
